src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Editor/TempFileTextBufferManager.cs

```diff
@@ -0,0 +1,181 @@
+        private readonly IProjectThreadingService _threadingService;
+        private readonly IServiceProvider _serviceProvider;
+
+        private IVsTextBuffer _textBuffer;
+        private ITextDocument _textDocument;
+        private IVsPersistDocData _docData;
+
+        [ImportingConstructor]
+        public TempFileTextBufferManager(UnconfiguredProject unconfiguredProject,
+            IProjectXmlAccessor msbuildAccessor,
+            IVsEditorAdaptersFactoryService editorAdaptersService,
+            ITextDocumentFactoryService textDocumentService,
+            IVsShellUtilitiesHelper shellUtilities,
+            IFileSystem fileSystem,
+            IProjectThreadingService threadingService,
+            [Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider) :
+            base(threadingService != null ? threadingService.JoinableTaskContext : throw new ArgumentNullException(nameof(threadingService)))
+        {
+            Requires.NotNull(unconfiguredProject, nameof(unconfiguredProject));
+            Requires.NotNull(msbuildAccessor, nameof(msbuildAccessor));
+            Requires.NotNull(editorAdaptersService, nameof(editorAdaptersService));
+            Requires.NotNull(textDocumentService, nameof(textDocumentService));
+            Requires.NotNull(shellUtilities, nameof(shellUtilities));
+            Requires.NotNull(fileSystem, nameof(fileSystem));
+            Requires.NotNull(serviceProvider, nameof(serviceProvider));
+
+            _unconfiguredProject = unconfiguredProject;
+            _msbuildAccessor = msbuildAccessor;
+            _editorAdaptersService = editorAdaptersService;
+            _textDocumentService = textDocumentService;
+            _shellUtilities = shellUtilities;
+            _fileSystem = fileSystem;
+            _threadingService = threadingService;
+            _serviceProvider = serviceProvider;
+        }
+
+        public string FilePath { get; set; }
+
+        public Task InitializeBufferAsync() => InitializeAsync();
+
+        protected override async Task InitializeCoreAsync(CancellationToken cancellationToken)
+        {
+            FilePath = GetTempFileName(Path.GetFileName(_unconfiguredProject.FullPath));
+
+            var projectXml = await _msbuildAccessor.GetProjectXmlAsync().ConfigureAwait(false);
+            _fileSystem.WriteAllText(FilePath, projectXml, await _unconfiguredProject.GetFileEncodingAsync().ConfigureAwait(false));
+        }
+
+        protected override Task DisposeCoreAsync(bool initialized)
+        {
+            _fileSystem.RemoveDirectory(Path.GetDirectoryName(FilePath), true);
+            return Task.CompletedTask;
+        }
+
+        public async Task ResetBufferAsync()
+        {
+            var projectXml = await _msbuildAccessor.GetProjectXmlAsync().ConfigureAwait(false);
+
+            // We compare the text we want to write with the text currently in the buffer, ignoring whitespace. If they're
```

Comment:
Can we file a bug to track this? This thing is still on the UI thread, and it can't be cheap for large files.